twister: tests: kernel: timer_behavior: Extend with stats recording

golowanow opened the pull request:
Extend `kernel/timer/timer_behavior` test suite with additional logging of the tests statistics gathered for timer drift, variance,
etc. as JSON-formatted records to ease data collection and analysis.

These log records will be processed by the Twister Harness recording feature which captures and parses timer statistics from the log output, then composes it into twister.json and recording.csv files.

The Twister Harness recording feature, as well as Twister Pytest Harness, are also improved to facilitate this change, improve logging, and in general, to allow data fields, extracted from the log by a regular expression, to be treated as a semi-structured JSON data.

golowanow committed:
twister: recording: Allow JSON data fields

Extend Twister Harness recording feature to allow selected data fields,
extracted from the log by a regular expression, to be parsed into JSON
objects and eventually reported in `twister.json` as `recording` list
property of the test suite.

With this extension, log records can convey layered data structures
passed from a test image as summary results, traces, statistics, etc.

This extension also allows flexible recording structure: a test image
can output different types of data records incapsulated into a fixed
set of fields, so `recording.csv` file columns are respected, whereas
some of the columns keep strings with json-encoded semi-structured data.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow committed:
tests: kernel: timer: Fix log capture on external tool testing

Fix incomplete test log capture when the external tool is used
and Twister Pytest Harness script ends without reading output
from the last test case running on the device.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow commented:
example log from `kernel.timer.timer` on frdm_k64f:

```
***** delaying boot 500ms (per build configuration) *****
*** Booting Zephyr OS build v3.6.0-4284-g38a05e7415c4 (delayed boot 500ms) ***
Running TESTSUITE timer_jitter_drift
===================================================================
START - test_jitter_drift_timer_period
periodic timer behavior test using built-in restart mechanism
collecting time samples for approx 10 seconds
periodic timer samples gathered, calculating statistics
timer clock rate 120000000, kernel tick rate 10000
period duration statistics for 10000 samples (0 rollovers):
expected: 1000 us,       	120000.000000 cycles
min:      999.983333 us, 	119998 cycles
max:      1000.025000 us, 	120003 cycles
mean:     1000.000000 us, 	120000.000000 cycles
variance: 0.000417 us, 	6.000000 cycles
stddev:   0.020412 us, 	2.449490 cycles
timer start cycle 64129401, end cycle 1264129401,
total time 10000000.000000 us, expected time 10000000.000000 us,
expected time drift 0.000000 us, difference 0.000000 us
RECORD: {"testcase":"jitter_drift_timer", "mechanism":"builtin", "stats_count":10000, "rollovers":0, "mean_us":1000.000000, "mean_cycles":120000, "stddev_us":0.020412, "stddev_cycles":2, "var_us":0.000417, "var_cycles":6, "min_us":999.983333, "min_cycles":119998, "max_us":1000.025000, "max_cycles":120003, "timer_start_cycle": 64129401, "timer_end_cycle": 1264129401, "total_time_us":10000000.000000, "expected_total_time_us":10000000.000000, "expected_total_drift_us":0.000000, "total_drift_us":0.000000, "expected_period_cycles":120000, "expected_period_drift_us":0.000000, "sys_clock_hw_cycles_per_sec":120000000, "CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC":120000000, "CONFIG_SYS_CLOCK_TICKS_PER_SEC":10000, "CONFIG_TIMER_TEST_PERIOD":1000, "CONFIG_TIMER_TEST_SAMPLES":10000, "CONFIG_TIMER_TEST_MAX_STDDEV":10}
PASS - test_jitter_drift_timer_period in 10.285 seconds
===================================================================
START - test_jitter_drift_timer_startdelay
periodic timer behavior test using explicit start with delay
collecting time samples for approx 10 seconds
periodic timer samples gathered, calculating statistics
timer clock rate 120000000, kernel tick rate 10000
period duration statistics for 10000 samples (0 rollovers):
expected: 1000 us,       	120000.000000 cycles
min:      999.983333 us, 	119998 cycles
max:      1000.025000 us, 	120003 cycles
mean:     1000.000000 us, 	120000.000000 cycles
variance: 0.000417 us, 	6.000000 cycles
stddev:   0.020412 us, 	2.449490 cycles
timer start cycle 1300020729, end cycle 2500020729,
total time 10000000.000000 us, expected time 10000000.000000 us,
expected time drift 0.000000 us, difference 0.000000 us
RECORD: {"testcase":"jitter_drift_timer", "mechanism":"startdelay", "stats_count":10000, "rollovers":0, "mean_us":1000.000000, "mean_cycles":120000, "stddev_us":0.020412, "stddev_cycles":2, "var_us":0.000417, "var_cycles":6, "min_us":999.983333, "min_cycles":119998, "max_us":1000.025000, "max_cycles":120003, "timer_start_cycle": 1300020729, "timer_end_cycle": 2500020729, "total_time_us":10000000.000000, "expected_total_time_us":10000000.000000, "expected_total_drift_us":0.000000, "total_drift_us":0.000000, "expected_period_cycles":120000, "expected_period_drift_us":0.000000, "sys_clock_hw_cycles_per_sec":120000000, "CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC":120000000, "CONFIG_SYS_CLOCK_TICKS_PER_SEC":10000, "CONFIG_TIMER_TEST_PERIOD":1000, "CONFIG_TIMER_TEST_SAMPLES":10000, "CONFIG_TIMER_TEST_MAX_STDDEV":10}
PASS - test_jitter_drift_timer_startdelay in 10.285 seconds
===================================================================
TESTSUITE timer_jitter_drift succeeded
Running TESTSUITE timer_tick_train
===================================================================
START - test_one_tick_timer_train
Initializing 4 Timers, Tick Rate 10000Hz, Expecting 25000 callbacks in 11000 ms
Starting Timers with Skews
Faking busy work, remaining timers is 4, timer callbacks 1
Faking busy work, remaining timers is 4, timer callbacks 639
Faking busy work, remaining timers is 4, timer callbacks 1277
Faking busy work, remaining timers is 4, timer callbacks 1915
Faking busy work, remaining timers is 4, timer callbacks 2553
Faking busy work, remaining timers is 4, timer callbacks 3191
Faking busy work, remaining timers is 4, timer callbacks 3830
Faking busy work, remaining timers is 4, timer callbacks 4468
Faking busy work, remaining timers is 4, timer callbacks 5106
Faking busy work, remaining timers is 4, timer callbacks 5744
Faking busy work, remaining timers is 4, timer callbacks 6383
Faking busy work, remaining timers is 4, timer callbacks 7021
Faking busy work, remaining timers is 4, timer callbacks 7659
Faking busy work, remaining timers is 4, timer callbacks 8297
Faking busy work, remaining timers is 4, timer callbacks 8936
Faking busy work, remaining timers is 4, timer callbacks 9574
Faking busy work, remaining timers is 4, timer callbacks 10212
Faking busy work, remaining timers is 4, timer callbacks 10851
Faking busy work, remaining timers is 4, timer callbacks 11489
Faking busy work, remaining timers is 4, timer callbacks 12128
Faking busy work, remaining timers is 4, timer callbacks 12766
Faking busy work, remaining timers is 4, timer callbacks 13404
Faking busy work, remaining timers is 4, timer callbacks 14043
Faking busy work, remaining timers is 4, timer callbacks 14681
Faking busy work, remaining timers is 4, timer callbacks 15320
Faking busy work, remaining timers is 4, timer callbacks 15958
Faking busy work, remaining timers is 4, timer callbacks 16597
Faking busy work, remaining timers is 4, timer callbacks 17235
Faking busy work, remaining timers is 4, timer callbacks 17874
Faking busy work, remaining timers is 4, timer callbacks 18512
Faking busy work, remaining timers is 4, timer callbacks 19151
Faking busy work, remaining timers is 4, timer callbacks 19789
Faking busy work, remaining timers is 4, timer callbacks 20428
Faking busy work, remaining timers is 4, timer callbacks 21066
Faking busy work, remaining timers is 4, timer callbacks 21705
Faking busy work, remaining timers is 4, timer callbacks 22343
Faking busy work, remaining timers is 4, timer callbacks 22982
Faking busy work, remaining timers is 4, timer callbacks 23620
Faking busy work, remaining timers is 4, timer callbacks 24259
Faking busy work, remaining timers is 4, timer callbacks 24897
One Tick Timer Train Done, took 10213 ms, busy loop ran 40 times
Perfect delta 48000 cycles or 400 us
Timer 0 max delta 48147 cycles or 401 us, 0 late callbacks (0.0%)
RECORD: {"testcase":"one_tick_timer_train", "timer":0, "max_delta_cycles":48147, "max_delta_us":401, "late_callbacks":0, "perfect_delta_cycles":48000, "perfect_delta_us":400, "train_time_ms":10213, "busy_loops":40, "timers":4, "expected_callbacks":25000, "expected_time_ms":11000, "CONFIG_SYS_CLOCK_TICKS_PER_SEC":10000}
Timer 1 max delta 48147 cycles or 401 us, 0 late callbacks (0.0%)
RECORD: {"testcase":"one_tick_timer_train", "timer":1, "max_delta_cycles":48147, "max_delta_us":401, "late_callbacks":0, "perfect_delta_cycles":48000, "perfect_delta_us":400, "train_time_ms":10213, "busy_loops":40, "timers":4, "expected_callbacks":25000, "expected_time_ms":11000, "CONFIG_SYS_CLOCK_TICKS_PER_SEC":10000}
Timer 2 max delta 48177 cycles or 401 us, 0 late callbacks (0.0%)
RECORD: {"testcase":"one_tick_timer_train", "timer":2, "max_delta_cycles":48177, "max_delta_us":401, "late_callbacks":0, "perfect_delta_cycles":48000, "perfect_delta_us":400, "train_time_ms":10213, "busy_loops":40, "timers":4, "expected_callbacks":25000, "expected_time_ms":11000, "CONFIG_SYS_CLOCK_TICKS_PER_SEC":10000}
Timer 3 max delta 48142 cycles or 401 us, 0 late callbacks (0.0%)
RECORD: {"testcase":"one_tick_timer_train", "timer":3, "max_delta_cycles":48142, "max_delta_us":401, "late_callbacks":0, "perfect_delta_cycles":48000, "perfect_delta_us":400, "train_time_ms":10213, "busy_loops":40, "timers":4, "expected_callbacks":25000, "expected_time_ms":11000, "CONFIG_SYS_CLOCK_TICKS_PER_SEC":10000}
PASS - test_one_tick_timer_train in 10.368 seconds
===================================================================
TESTSUITE timer_tick_train succeeded
------ TESTSUITE SUMMARY START ------
SUITE PASS - 100.00% [timer_jitter_drift]: pass = 2, fail = 0, skip = 0, total = 2 duration = 20.570 seconds
- PASS - [timer_jitter_drift.test_jitter_drift_timer_period] duration = 10.285 seconds
- PASS - [timer_jitter_drift.test_jitter_drift_timer_startdelay] duration = 10.285 seconds
SUITE PASS - 100.00% [timer_tick_train]: pass = 1, fail = 0, skip = 0, total = 1 duration = 10.368 seconds
- PASS - [timer_tick_train.test_one_tick_timer_train] duration = 10.368 seconds
------ TESTSUITE SUMMARY END ------
===================================================================
RunID: ccd10bd39d1a18ec646eeff092d5004d
PROJECT EXECUTION SUCCESSFUL

```

golowanow commented:
example Twister+Pytest log from `kernel.timer.timer_behavior_external` on frdm_k64f with a logic analyzer connected, so there are 'RECORD' rows from both the test image and external measurements:

```
PYTEST: DEBUG: #: *** Booting Zephyr OS build v3.6.0-4284-g38a05e7415c4 (delayed boot 500ms) ***
PYTEST: DEBUG: #: Running TESTSUITE timer_jitter_drift
PYTEST: DEBUG: #: ===================================================================
PYTEST: DEBUG: #: START - test_jitter_drift_timer_period
PYTEST: DEBUG: #: periodic timer behavior test using built-in restart mechanism
PYTEST: DEBUG: #: ===== External Tool Sync Point =====
PYTEST: DEBUG: #: ===== builtin =====
PYTEST: INFO: sub ChannelConnectivity.IDLE
PYTEST: INFO: sub ChannelConnectivity.CONNECTING
PYTEST: INFO: sub ChannelConnectivity.READY
PYTEST: INFO: ====================================================
PYTEST: INFO: periodic timer behaviour using builtin mechanism:
PYTEST: INFO: min:            1000.250000 us
PYTEST: INFO: max:            1000.375000 us
PYTEST: INFO: mean:           1000.343050 us
PYTEST: INFO: variance:       0.000000 us
PYTEST: INFO: stddev:         0.054525 us
PYTEST: INFO: total time:     10003430.500000 us
PYTEST: INFO: expected drift: 45000.0 us
PYTEST: INFO: real drift:     3430.500000 us
PYTEST: INFO: ====================================================
PYTEST: INFO: RECORD: {"testcase":"jitter_drift_timer", "mechanism":"builtin_external", "stats_count": 10000, "mean_us":1000.343050, "stddev_us":0.054525, "var_us":0.000000, "min_us":1000.250000, "max_us":1000.375000, "total_time_us":10003430.500000, "expected_total_time_us":10000000.000000, "expected_total_drift_us":45000.000000, "total_drift_us": 3430.500000, "min_bound_us":900.000000, "max_bound_us":1100.000000, "expected_period_cycles":120000, "sys_clock_hw_cycles_per_sec":120000000, "CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC":120000000, "CONFIG_SYS_CLOCK_TICKS_PER_SEC":10000, "CONFIG_TIMER_TEST_PERIOD":1000, "CONFIG_TIMER_TEST_SAMPLES":10000, "CONFIG_TIMER_TEST_MAX_STDDEV":10, "CONFIG_TIMER_EXTERNAL_TEST_PERIOD_MAX_DRIFT_PPM":100000, "CONFIG_TIMER_EXTERNAL_TEST_MAX_DRIFT_PPM":4500}
PYTEST: DEBUG: #: ===== Waiting 3 seconds =====
PYTEST: DEBUG: #: collecting time samples for approx 10 seconds
PYTEST: DEBUG: #: periodic timer samples gathered, calculating statistics
PYTEST: DEBUG: #: timer clock rate 120000000, kernel tick rate 10000
PYTEST: DEBUG: #: period duration statistics for 10000 samples (0 rollovers):
PYTEST: DEBUG: #: expected: 1000 us,       	120000.000000 cycles
PYTEST: DEBUG: #: min:      999.966667 us, 	119996 cycles
PYTEST: DEBUG: #: max:      1000.008333 us, 	120001 cycles
PYTEST: DEBUG: #: mean:     1000.000000 us, 	120000.000000 cycles
PYTEST: DEBUG: #: variance: 0.000278 us, 	4.000000 cycles
PYTEST: DEBUG: #: stddev:   0.016667 us, 	2.000000 cycles
PYTEST: DEBUG: #: timer start cycle 425053443, end cycle 1625053443,
PYTEST: DEBUG: #: total time 10000000.000000 us, expected time 10000000.000000 us,
PYTEST: DEBUG: #: expected time drift 0.000000 us, difference 0.000000 us
PYTEST: DEBUG: #: RECORD: {"testcase":"jitter_drift_timer", "mechanism":"builtin", "stats_count":10000, "rollovers":0, "mean_us":1000.000000, "mean_cycles":120000, "stddev_us":0.016667, "stddev_cycles":2, "var_us":0.000278, "var_cycles":4, "min_us":999.966667, "min_cycles":119996, "max_us":1000.008333, "max_cycles":120001, "timer_start_cycle": 425053443, "timer_end_cycle": 1625053443, "total_time_us":10000000.000000, "expected_total_time_us":10000000.000000, "expected_total_drift_us":0.000000, "total_drift_us":0.000000, "expected_period_cycles":120000, "expected_period_drift_us":0.000000, "sys_clock_hw_cycles_per_sec":120000000, "CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC":120000000, "CONFIG_SYS_CLOCK_TICKS_PER_SEC":10000, "CONFIG_TIMER_TEST_PERIOD":1000, "CONFIG_TIMER_TEST_SAMPLES":10000, "CONFIG_TIMER_TEST_MAX_STDDEV":10}
PYTEST: DEBUG: #: PASS - test_jitter_drift_timer_period in 13.291 seconds
PYTEST: DEBUG: #: ===================================================================
PYTEST: DEBUG: #: START - test_jitter_drift_timer_startdelay
PYTEST: DEBUG: #: periodic timer behavior test using explicit start with delay
PYTEST: DEBUG: #: ===== External Tool Sync Point =====
PYTEST: DEBUG: #: ===== startdelay =====
PYTEST: INFO: sub ChannelConnectivity.IDLE
PYTEST: INFO: sub ChannelConnectivity.CONNECTING
PYTEST: INFO: sub ChannelConnectivity.READY
PYTEST: INFO: ====================================================
PYTEST: INFO: periodic timer behaviour using startdelay mechanism:
PYTEST: INFO: min:            1000.250000 us
PYTEST: INFO: max:            1000.375000 us
PYTEST: INFO: mean:           1000.343025 us
PYTEST: INFO: variance:       0.000000 us
PYTEST: INFO: stddev:         0.054539 us
PYTEST: INFO: total time:     10003430.250000 us
PYTEST: INFO: expected drift: 45000.0 us
PYTEST: INFO: real drift:     3430.250000 us
PYTEST: INFO: ====================================================
PYTEST: INFO: RECORD: {"testcase":"jitter_drift_timer", "mechanism":"startdelay_external", "stats_count": 10000, "mean_us":1000.343025, "stddev_us":0.054539, "var_us":0.000000, "min_us":1000.250000, "max_us":1000.375000, "total_time_us":10003430.250000, "expected_total_time_us":10000000.000000, "expected_total_drift_us":45000.000000, "total_drift_us": 3430.250000, "min_bound_us":900.000000, "max_bound_us":1100.000000, "expected_period_cycles":120000, "sys_clock_hw_cycles_per_sec":120000000, "CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC":120000000, "CONFIG_SYS_CLOCK_TICKS_PER_SEC":10000, "CONFIG_TIMER_TEST_PERIOD":1000, "CONFIG_TIMER_TEST_SAMPLES":10000, "CONFIG_TIMER_TEST_MAX_STDDEV":10, "CONFIG_TIMER_EXTERNAL_TEST_PERIOD_MAX_DRIFT_PPM":100000, "CONFIG_TIMER_EXTERNAL_TEST_MAX_DRIFT_PPM":4500}
PYTEST: DEBUG: #: ===== Waiting 3 seconds =====
PYTEST: DEBUG: #: collecting time samples for approx 10 seconds
PYTEST: DEBUG: #: periodic timer samples gathered, calculating statistics
PYTEST: DEBUG: #: timer clock rate 120000000, kernel tick rate 10000
PYTEST: DEBUG: #: period duration statistics for 10000 samples (0 rollovers):
PYTEST: DEBUG: #: expected: 1000 us,       	120000.000000 cycles
PYTEST: DEBUG: #: min:      999.966667 us, 	119996 cycles
PYTEST: DEBUG: #: max:      1000.008333 us, 	120001 cycles
PYTEST: DEBUG: #: mean:     1000.000000 us, 	120000.000000 cycles
PYTEST: DEBUG: #: variance: 0.000278 us, 	4.000000 cycles
PYTEST: DEBUG: #: stddev:   0.016667 us, 	2.000000 cycles
PYTEST: DEBUG: #: timer start cycle 2021748850, end cycle 3221748850,
PYTEST: DEBUG: #: total time 10000000.000000 us, expected time 10000000.000000 us,
PYTEST: DEBUG: #: expected time drift 0.000000 us, difference 0.000000 us
PYTEST: DEBUG: #: RECORD: {"testcase":"jitter_drift_timer", "mechanism":"startdelay", "stats_count":10000, "rollovers":0, "mean_us":1000.000000, "mean_cycles":120000, "stddev_us":0.016667, "stddev_cycles":2, "var_us":0.000278, "var_cycles":4, "min_us":999.966667, "min_cycles":119996, "max_us":1000.008333, "max_cycles":120001, "timer_start_cycle": 2021748850, "timer_end_cycle": 3221748850, "total_time_us":10000000.000000, "expected_total_time_us":10000000.000000, "expected_total_drift_us":0.000000, "total_drift_us":0.000000, "expected_period_cycles":120000, "expected_period_drift_us":0.000000, "sys_clock_hw_cycles_per_sec":120000000, "CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC":120000000, "CONFIG_SYS_CLOCK_TICKS_PER_SEC":10000, "CONFIG_TIMER_TEST_PERIOD":1000, "CONFIG_TIMER_TEST_SAMPLES":10000, "CONFIG_TIMER_TEST_MAX_STDDEV":10}
PYTEST: DEBUG: #: PASS - test_jitter_drift_timer_startdelay in 13.292 seconds
PYTEST: DEBUG: #: ===================================================================
PYTEST: DEBUG: #: TESTSUITE timer_jitter_drift succeeded
PYTEST: DEBUG: #: ------ TESTSUITE SUMMARY START ------
PYTEST: DEBUG: #: SUITE PASS - 100.00% [timer_jitter_drift]: pass = 2, fail = 0, skip = 0, total = 2 duration = 26.583 seconds
PYTEST: DEBUG: #: - PASS - [timer_jitter_drift.test_jitter_drift_timer_period] duration = 13.291 seconds
PYTEST: DEBUG: #: - PASS - [timer_jitter_drift.test_jitter_drift_timer_startdelay] duration = 13.292 seconds
PYTEST: DEBUG: #: SUITE SKIP -   0.00% [timer_tick_train]: pass = 0, fail = 0, skip = 1, total = 1 duration = 0.000 seconds
PYTEST: DEBUG: #: - SKIP - [timer_tick_train.test_one_tick_timer_train] duration = 0.000 seconds
PYTEST: DEBUG: #: ------ TESTSUITE SUMMARY END ------
PYTEST: DEBUG: #: ===================================================================
PYTEST: DEBUG: #: RunID: be823331b69ea4db8519bbafaeda4242
PYTEST: DEBUG: #: PROJECT EXECUTION SUCCESSFUL
PYTEST: PASSED
```

golowanow commented:
an example snippet from `twister.json`:

```
        {
            "name":"tests/kernel/timer/timer_behavior/kernel.timer.timer",
            "arch":"arm",
            "platform":"frdm_k64f",
            "path":"tests/kernel/timer/timer_behavior",
            "run_id":"ccd10bd39d1a18ec646eeff092d5004d",
            "runnable":true,
            "retries":0,
            "dut":"02400b0147784e4500029004d7450023ddb1000097969900",
            "status":"passed",
            "execution_time":"35.79",
            "build_time":"8.70",
            "testcases":[
                {
                    "identifier":"kernel.timer.timer.jitter_drift_timer_startdelay",
                    "execution_time":"10.29",
                    "status":"passed"
                },
                {
                    "identifier":"kernel.timer.timer.jitter_drift_timer_period",
                    "execution_time":"10.29",
                    "status":"passed"
                },
                {
                    "identifier":"kernel.timer.timer.one_tick_timer_train",
                    "execution_time":"10.37",
                    "status":"passed"
                }
            ],
            "recording":[
                {
                    "metrics":{
                        "testcase":"jitter_drift_timer",
                        "mechanism":"builtin",
                        "stats_count":10000,
                        "rollovers":0,
                        "mean_us":1000.0,
                        "mean_cycles":120000,
                        "stddev_us":0.020412,
                        "stddev_cycles":2,
                        "var_us":0.000417,
                        "var_cycles":6,
                        "min_us":999.983333,
                        "min_cycles":119998,
                        "max_us":1000.025,
                        "max_cycles":120003,
                        "timer_start_cycle":64129401,
                        "timer_end_cycle":1264129401,
                        "total_time_us":10000000.0,
                        "expected_total_time_us":10000000.0,
                        "expected_total_drift_us":0.0,
                        "total_drift_us":0.0,
                        "expected_period_cycles":120000,
                        "expected_period_drift_us":0.0,
                        "sys_clock_hw_cycles_per_sec":120000000,
                        "CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC":120000000,
                        "CONFIG_SYS_CLOCK_TICKS_PER_SEC":10000,
                        "CONFIG_TIMER_TEST_PERIOD":1000,
                        "CONFIG_TIMER_TEST_SAMPLES":10000,
                        "CONFIG_TIMER_TEST_MAX_STDDEV":10
                    }
                },
.....
```

andyross commented on scripts/schemas/twister/testsuite-schema.yaml:
```diff
@@ -130,6 +130,11 @@ schema;scenario-schema:
             "regex":
               type: str
               required: true
+            "as_json":
```
Not enough of a twister expert to really review here, but it would be good to add a section to the docs somewhere describing this and giving a simple example of usage.

golowanow replied:
right, documentation added.

andyross commented on tests/kernel/timer/timer_behavior/pytest/test_timer.py:
```diff
@@ -99,3 +99,6 @@ def test_flash(dut: DeviceAdapter, tool, tool_options, config,
         do_analysys(test, stats, stats_count, config, sys_clock_hw_cycles_per_sec)
+
+    # Let the running test's image output to be fully captured from device.
+    dut.readlines()
```
This relies on an EOF condition on the stream, no?  What about things like UARTs that say open forever?  Or is there handling at another layer for detecting "end of test"?

edersondisouza replied:
I'm kinda confused why is this needed at all: when running with the external tool, the output of the Zephyr application shouldn't matter at all, what matters is what was collected from the external tool.

golowanow replied:
> This relies on an EOF condition on the stream, no? What about things like UARTs that say open forever? Or is there handling at another layer for detecting "end of test"?

in this case, it is either Twister Pytest `DeviceAdapter` and `Harness` responsibility, or the board-related setup also at the Twister layer to release the board on timeout.

golowanow replied:
> I'm kinda confused why is this needed at all: when running with the external tool, the output of the Zephyr application shouldn't matter at all, what matters is what was collected from the external tool.

full Zephyr application test output matters for two reasons :
1) to have evidence the test is actually passed and the external data is valid;
2) to have the test app's statistics collected as well, which is interesting to analyze for history stats, and to compare with the timer test app variant which runs without the external tool on the sane board.

edersondisouza commented on tests/kernel/timer/timer_behavior/src/jitter_drift.c:
Maybe separate the change to the C files on a different patch, so it's clear what is needed for the external tool test and what is for the in-app only?

golowanow replied:
done, now in separate commits

edersondisouza reviewed (commented):
Looking good. I just think there should be some separation per tests here: a patch that adds the stats output for the in-app tests and another that touches the external tool side.

golowanow committed:
tests: kernel: timer_behavior: Additional stats recording

Additional logging of kernel.timer.timer test case statistics for
timer drift, variance, etc. as JSON-formatted records to make easier
data collection and its further analysis.
These log records will be processed by the Twister Harness recording
feature which captures and parses timer statistics from the log output,
then composes it into twister.json and recording.csv files.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow committed:
tests: kernel: timer_behavior_external: Additional stats recording

Additional logging of kernel.timer.timer_behavior_external test case
statistics for timer drift, variance, etc. as JSON-formatted records
to make easier data collection and its further analysis.
These log records will be processed by the Twister Harness recording
feature which captures and parses timer statistics from the log output,
then composes it into twister.json and recording.csv files.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow commented on tests/kernel/timer/timer_behavior/README:
```diff
@@ -42,6 +42,8 @@ following statistics, in seconds:
     'max':          Maximum period registered
     'total_time':   Total time, between first and last period.
 
+and an integer value of how many data values are collected for the statistics.
```
> Any reason why couldn't the size of the set be inside the dictionary?

as the dictionary keeps time statistics only, as it was before this PR, and it is also easier to deal with e.g. when converting to microseconds and formatting output.

hakehuang commented:
@golowanow is it possible to support simulator with this application, and using a script to output the expected timing?

golowanow commented:
> @golowanow is it possible to support simulator with this application, and using a script to output the expected timing?

The proposed extension in the timer_behavior test should work for any platform which executes it.
The twister extension is also generic, and it is in the same PR just to facilitate the test change.

golowanow commented:
dear reviewers, do you have any comments or objections ?

LukaszMrugala commented on tests/kernel/timer/timer_behavior/pytest/test_timer.py:
```diff
@@ -11,7 +11,7 @@
 
 
-def do_analysys(test, stats, config, sys_clock_hw_cycles_per_sec):
+def do_analysys(test, stats, stats_count, config, sys_clock_hw_cycles_per_sec):
```
Typo: `do_analysis`, no?

golowanow replied:
good catch, lets fix it.

golowanow commented:
@PerMac, @gopiotr, @gchwier, @hakehuang - ping: any more notes on the twister part of this PR ?

PerMac reviewed (commented):
Why do you want to have such metrics in twister.json? From my understanding, this data will have quite a custom form in the json report, depending on test yaml. I mean that the entry can be called "metrics" as in the example but can also have any other name. Having this in mind: what benefit will bring having such data in the twister.json report? I guess one would need also a custom parser to get these data for any usage? Wouldn't it be better to store such recording as an independent file?

golowanow commented:
> Having this in mind: what benefit will bring having such data in the twister.json report?

We can import these statistics into ElasticSearch to analyze trends or unexpected changes of timer statistics.

> Wouldn't it be better to store such recording as an independent file?

It is more reliable and easier to extract data from twister.json with all its context of when/how it was obtained, using tools like jq.

gchwier commented on tests/kernel/timer/timer_behavior/pytest/test_timer.py:
```diff
@@ -52,6 +52,29 @@ def do_analysys(test, stats, config, sys_clock_hw_cycles_per_sec):
     logger.info('====================================================')
 
+    logger.info('RECORD: {'
+                f'"testcase":"jitter_drift_timer"'
```
would be easier to create json with json lib and use just json.dumps method

golowanow replied:
isn't it much easier just to print ? there is no need to parse json here.

golowanow replied:
@gchwier, changed from print to json.dumps

gchwier commented on tests/kernel/timer/timer_behavior/pytest/test_timer.py:
```diff
@@ -52,6 +52,29 @@ def do_analysys(test, stats, config, sys_clock_hw_cycles_per_sec):
     logger.info(f'real drift:     {time_diff * 1_000_000:.6f} us')
     logger.info('====================================================')
 
+    logger.info('RECORD: {'
```
`logger.info` will not be printed by twister, if you use verbose less than 2,
see: https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/pylib/twister/twisterlib/harness.py#L344
```
        if handler.options.verbose > 1:
            command.extend([
                '--log-cli-level=DEBUG',
                '--log-cli-format=%(levelname)s: %(message)s'
            ])
```
and then in `_ourtput_reader` method tou do not have any line to parse:
```
        while proc.stdout.readable() and proc.poll() is None:
            line = proc.stdout.readline().decode().strip()
            if not line:
                continue
            logger.debug('PYTEST: %s', line)
            harness.parse_record(line)
```

Test it with `twister -v`, no metrics in twister.json

golowanow replied:
Thank you for pointing at this: the 'external' test will have no output logged following the Twister verbose level.
I've added logging options for pytest plugin in the testcase.yaml explicitly.

gchwier replied:
It is not good solution to add it in testcase.yaml. You must then remember to update pytest-args in every sample/testcase yaml when you want to use such a metrics.
I think that now there is no reason to remove `--log-cli-level=DEBUG`, when verbosity is lower in twister (Twister does not print it).
So better to have generic solution by adding:
```
                '--log-cli-level=DEBUG',
                '--log-cli-format=%(levelname)s: %(message)s'
```
to command always.
Another benefit from it would be to have all debug logs in the twister output when pytest fails (with twister `--inline-logs`)

golowanow replied:
> It is not good solution to add it in testcase.yaml. .... So better to have generic solution by adding: ...
> to command always. Another benefit from it would be to have all debug logs in the twister output when pytest fails (with twister `--inline-logs`)

I agree, it is a general pytest harness problem: to have both outputs (from the test image and from the  pytest test) to pass through pytest plugin into Twister logs.

golowanow committed:
tests: kernel: timer_behavior_external: Fix a typo

Rename `do_analysys()` to `do_analysis()`.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow commented on tests/kernel/timer/timer_behavior/testcase.yaml:
```diff
@@ -8,13 +8,22 @@ tests:
     harness: pytest
     harness_config:
-      pytest_args: ['--tool', 'saleae_logic2', '--tool-options',
+      pytest_args: ['--log-cli-level', 'DEBUG',
+                    '--log-cli-format', '%(levelname)s: %(message)s',
```
ok, but isn't it a bit out of scope already ? I need some time to check how it works.

gchwier replied:
Maybe a bit out out of scope, but it is easy change and will fix problem of gathering metrics when verbosity is not enabled (and also problem with printing a debug output when pytest scenario fails).
Adding pytest_args to every sample/testcase yaml to gather metrics is not convenient workaround

golowanow replied:
added as suggested

golowanow committed:
twister: pytest: Log output passthrough

Pass Twister pytest plugin's log output, as well as output from
a test image running with pytest, up to Twister log irregardless
of the current verbosity level set at Twister.
This allows to collect full test output and also ensures that
recording data embedded in the test log is passed to the Twister
pytest harness for export.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow commented:
@andyross, @edersondisouza, @hakehuang, @LukaszMrugala, @ceolin, @PerMac, @kartben, 
please have a look again / refresh your reviews.
